### ISSUE TITLE
Relation controller interface in backend forms doesn't work with a belongsTo relationship if the primary / master record hasn't been created first

### ISSUE THREAD

**Issue by lthurston** [octobercms/october#1329]
I have a fork of the oc-test-plugin that serves as a proof of concept. I had to make some modifications to it because this issue was obscured by virtue of which views the relation manager was set up in. Note that I removed some validation on name fields to get to the heart of the issue. With the validation enabled, the bug doesn't occur, but that's only because the primary / master record cannot be validated nor saved.

My fork is here: https://github.com/lthurston/oc-test-plugin

The readme provides a bug summary and instructions to experience the issue yourself.

Expected result: ideally, using the relation controller in unsaved records would work, and the unsaved master record would be created only once the record was saved by clicking "save" and not before.

Actual result: saving the slave record results in the creation of a blank master record with only the relationship field set. When the editor saves the record with their details, a second record gets created, without the relationship.

**Comment by lthurston** [comment]
I wanted to follow up with some thoughts on the solution that I've provided here via this pair of pull requests. The core problem isn't so much that belongsTo relation and deferred bindings don't work, but more that the relationship controller will create the master record when the slave is saved. If the master record form itself created the master record along with the belongsTo relation to the slave when the user clicked "save", then this wouldn't be a problem.

It's a little counter-intuitive that deferred bindings would be needed on belongsTo relationships. The slave record is the first to be saved, and since the relation to that slave is on the master, why would you need to defer the binding at all? In most cases you probably don't. But when there's a degree of separation between the creation of the master and slave -- like in this exact case of the relationship manager interface -- deferred binding actually feels like an elegant solution. This situation is a case in point of the reason why disallowing deferred binding on belongsTo doesn't make a ton of sense. It's a edge case, but it's still a real case.

That said, this isn't the only solution to the problem. The other obvious solution would be to not save the master record when the slave is saved but instead save from the form only. This would likely require some conditional logic in the relationship manager and the master form. I haven't dug into this.

Although the changes that I provided in the pull requests may not be the perfect solution, I'm not sure the other solution I just mentioned is in any way better.

As always, I'm open to suggestions. But one way or another saving belongsTo relations via the relation manager should be possible in cases when the master record has already been saved and when it hasn't. The solution that allows the unusual case of deferred bindings for belongsTo feels like the right one to me. I'm happy to modify my changes in any way to get them into the core.

**Comment by lthurston** [comment]
I've added another pull request with a simple change that reloads relations before generating the relation controller preview partial. As it describes, this forces relations to be pulled in so they can be previewed (like an attachOne image).

**Comment by daftspunk** [comment]
Upon testing it seems like the issue only occurs when the Phone is being created, not when a Phone is being updated. This process shouldn't require any Deferred Binding because the Person already exists at the time the Phone is being created, so `$phone->person_id` can easily be set at this point.

It seems like a logic error to me.

**Comment by daftspunk** [comment]
To prove my assertion above, these instructions also produce the same error

1. [...]
1. Under "Alternate Views" click "Relation Controller"
1. Click "Link Person"
1. Select an existing person
1. Add a name to the phone and save
1. [...]

**Comment by lthurston** [comment]
I don't disagree; this is only a problem when the phone doesn't exist. You're suggesting the fix should happen in the relation manager though, rather than in a deferred binding-based solution?

As I tried to clarify above, although there's no technical reason per se that a belongs to should need deferred binding, if there are other technical reasons that prevent the master record from knowing about the slave record then it becomes useful. That's the case here with this interface.

This leads me to think that allowing deferred binding on belongs to -- despite being not necessary in most cases -- isn't necessary something you want to prevent from happening at all.

**Comment by daftspunk** [comment]
Still in discovery mode, will get back to you soon. However at this point I think you have proven there is a case for belongsTo to be deferred.

**Comment by daftspunk** [comment]
The problem we have here is RelationController has no idea about FormController, if it did, we could simply add this to the preview form

    <input type="hidden" name="Phone[person_id]" value="<person id I just created/linked>" />

And the form will pick that up and use it. RelationController should be abstract from the FormControllers expectations, what if we were using some other implementation? It has to work for that too. So it would make sense that Deferred Binding is the way to go, perhaps.

**Comment by lthurston** [comment]
Exactly. I think we're seeing eye to eye on this.

**Comment by daftspunk** [comment]
Ok - so assume that this the way forward. The belongsTo relationship is unique because the commit must occur *before* the model is saved, whereas all the others occur afterwards.

1. Master model saves
1. Deferred query for slaves
1. Slaves are saved against newly created model

In the logic above the master model would get saved twice, once by the called save, and once by the deferred commit process.

With this new logic we need

1. Before model saves
1. Deferred query for slaves
1. BelongsTo slaves are "set" against the master model
1. Master model saves
1. All other Slaves are saved against newly created model

It complicates things further, the deferred query needs to be cached. But that's alright I suppose. I will keep looking at it.

**Comment by daftspunk** [comment]
Wow - this is a tough call, I tend to agree. For an outsider looking in, the solution we add to RelationController would be far more confusing to comprehend than the solution we add to Deferred Binding in models. The latter also adds more value in general - if this scenario were to repeat elsewhere for someone else. It's also more consistent from an API perspective to allow ->add()/remove() to belongsTo relationships. 

Well that's 3 strong points for improving deferred binding. Heh, I never thought I'd see the day =)

**Comment by daftspunk** [comment]
All sorted

**Comment by lthurston** [comment]
Thanks, I appreciate your help.
